subsys/scsi/scsi.c

```diff
@@ -0,0 +1,513 @@
+	struct scsi_cmd cmd = {0};
+
+	/* Set up the SCSI command structure for TUR */
+	cmd.lun = (uint8_t)lun;
+	cmd.datalen = (0);
+	cmd.dma_dir = (uint8_t)PERIPHERAL_TO_PERIPHERAL;
+
+	/* Setup the "Test Unit Ready" command */
+	scsi_setup_test_unit_ready(&cmd);
+
+	/* Execute the SCSI command and return the result */
+	do {
+		ret = scsi_exec(scsi_dev, &cmd);
+		retries--;
+	} while ((ret < 0) && (retries > 0));
+
+	return ret;
+}
+
+/**
+ * @brief Perform a write operation on the SCSI device.
+ *
+ * This function sends a write command to the SCSI device.
+ *
+ * @param scsi_dev Pointer to the scsi_device structure.
+ * @param sector Starting sector number.
+ * @param count Number of sectors to write.
+ * @param buf Pointer to the data buffer to write.
+ *
+ * @return 0 if successful, or a negative error code on failure.
+ */
+int32_t scsi_write(struct scsi_device *scsi_dev, uint64_t sector,
+		   uint32_t count, const uint8_t *buf)
+{
+	int32_t ret;
+	struct scsi_cmd cmd = {0};
+
+	if (scsi_dev == NULL) {
+		return -EINVAL;
+	}
+
+	/* Set up the SCSI command structure */
+	cmd.lun = scsi_dev->lun;
+	cmd.pdata = (uint8_t *)(uintptr_t)buf;
+	cmd.datalen = ((uint64_t)(scsi_dev->sector_size) * (uint64_t)(count));
+	cmd.dma_dir = (uint8_t)MEMORY_TO_PERIPHERAL;
+
+	/* Setup the "Write" command with specified sector and count */
+	scsi_setup_write(&cmd, sector, count);
+
+	/* Execute the SCSI command and return the result */
+	ret = scsi_exec(scsi_dev, &cmd);
+
+	return ret;
+}
+
+/**
+ * @brief Perform a read operation on the SCSI device.
+ *
+ * This function sends a read command to the SCSI device.
+ *
+ * @param scsi_dev Pointer to the scsi_device structure.
+ * @param sector Starting sector number.
+ * @param count Number of sectors to read.
+ * @param buf Pointer to the buffer to store the read data.
+ *
+ * @return 0 if successful, or a negative error code on failure.
+ */
+int32_t scsi_read(struct scsi_device *scsi_dev, uint64_t sector,
+		  uint32_t count, uint8_t *buf)
+{
+	int32_t ret;
+	struct scsi_cmd cmd = {0};
+
+	if (scsi_dev == NULL) {
+		return -EINVAL;
+	}
+
+	/* Set up the SCSI command structure */
+	cmd.lun = scsi_dev->lun;
+	cmd.pdata = (uint8_t *)buf;
+	cmd.datalen = ((uint64_t)(scsi_dev->sector_size) * (uint64_t)(count));
+	cmd.dma_dir = (uint8_t)PERIPHERAL_TO_MEMORY;
+
+	/* Setup the "Read" command with specified sector and count */
+	scsi_setup_read(&cmd, sector, count);
+
+	/* Execute the SCSI command and return the result */
+	ret = scsi_exec(scsi_dev, &cmd);
+
+	return ret;
+}
+
+/**
+ * @brief Handle SG_IO ioctl for SCSI device.
+ *
+ * This function handles the SG_IO ioctl command for the SCSI device,
+ * which allows for sending arbitrary SCSI commands.
+ *
+ * @param sdev Pointer to the scsi_device structure.
+ * @param arg Pointer to the ioctl arguments.
+ *
+ * @return 0 if successful, or a negative error code on failure.
+ */
+static int32_t scsi_ioctl_sg_io(struct scsi_device *sdev, void *arg)
```

Comment:
Why use a `void *` here, over just using the `sq_io_req` pointer as part of the function signature?